OpenRA.Mods.Common/Traits/World/WeatherOverlay.cs

```diff
@@ -0,0 +1,391 @@
+		{
+			public float PosX;
+			public float PosY;
+			public int Size;
+			public float DirectionScatterX;
+			public float Gravity;
+			public float SwingOffset;
+			public float SwingSpeed;
+			public bool SwingDirection;
+			public float SwingAmplitude;
+			public Color Color;
+			public Color TailColor;
+
+			public Particle(
+				float posX,
+				float posY,
+				int size,
+				float directionScatterX,
+				float gravity,
+				float swingOffset,
+				float swingSpeed,
+				bool swingDirection,
+				float swingAmplitude,
+				Color color,
+				Color tailColor)
+			{
+				PosX = posX;
+				PosY = posY;
+				Size = size;
+				DirectionScatterX = directionScatterX;
+				Gravity = gravity;
+				SwingOffset = swingOffset;
+				SwingSpeed = swingSpeed;
+				SwingDirection = swingDirection;
+				SwingAmplitude = swingAmplitude;
+				Color = color;
+				TailColor = tailColor;
+			}
+		}
+
+		public List<Particle> ParticleList = new List<Particle>();
+		int maxParticleCount;
+
+		enum ParticleCountFaderType { Hold, FadeIn, FadeOut }
+		ParticleCountFaderType particleCountFader = ParticleCountFaderType.FadeIn;
+
+		float targetWindXOffset = 0f;
+		float currentWindXOffset = 0f;
+		int currentWindIndex = 0;
+		long windTickCountdown = 1500;
+		float2 antiScrollPrevTopLeft;
+
+		public WeatherOverlay(World world, WeatherOverlayInfo info)
+		{
+			this.info = info;
+			this.world = world;
+			currentWindIndex = info.WindStrength.Length / 2;
+			targetWindXOffset = info.WindStrength[0];
+			maxParticleCount = CalculateParticleCount(Game.Renderer.Resolution.Width, Game.Renderer.Resolution.Height);
+		}
+
+		int CalculateParticleCount(int x, int y)
+		{
+			var tempfactor = 0;
+			if (info.ParticleDensityFactor != 0)
+				tempfactor = info.ParticleDensityFactor;
+			else tempfactor = 1;
+
+			return (x * y) / tempfactor;
+		}
+
+		void SpawnParticles(int count, int rangeY, int spawnChancePercent)
+		{
+			for (int i = 0; i < count; i++)
+			{
+				if (Game.CosmeticRandom.Next(0, 101) <= spawnChancePercent)
+				{
+					var tempColor = info.ParticleColors[Game.CosmeticRandom.Next(0, info.ParticleColors.Length / 4 + 1)];
+					var tempColorTail = Color.FromArgb(info.LineTailAlphaValue, tempColor.R, tempColor.G, tempColor.B);
+					var tempSwingDirection = Game.CosmeticRandom.Next(0, 2) == 0 ? true : false;
+
+					ParticleList.Add(
+						new Particle(
+							Game.CosmeticRandom.Next(0, Game.Renderer.Resolution.Width + 1),
+							Game.CosmeticRandom.Next(0, rangeY + 1),
+							Game.CosmeticRandom.Next(info.PaticleSize[0], info.PaticleSize[1] + 1),
+							info.DirectionScatter[0] + Game.CosmeticRandom.Next(0, info.DirectionScatter[1] - info.DirectionScatter[0] + 1),
+							(info.Gravity[0] + Game.CosmeticRandom.Next(0, info.Gravity[1] - info.Gravity[0] + 1)) / 100f,
+							Game.CosmeticRandom.Next(info.SwingOffset[0], info.SwingOffset[1] + 1) / 10f,
+							Game.CosmeticRandom.Next(info.SwingSpeed[0], info.SwingSpeed[1] + 1) / 1000f,
```

Comment:
That Game.CosmeticRandom.Next() does not support float. -> thats how i get my floats